### PR TITLE
feat: permit build to directly receive files to be assembled

### PR DESCRIPTION
**Pull request by BretasArthur1**
## Problem

The command `build` didn't had an option to directly supply files to be assembled.

## Changes
- Add a new flag `input` to `BuildArgs` struct to directly receive a file to be assembled.
- Introduces new `compile_result` fn to print the compilation status.
- Refactor the old `build` function removing internal functions and making them standalone. 

## Testing

Compiled and tested files supplied directly with the new argument.

### CODE REVIEW

**Comment by BretasArthur1** [comment]
Screenshot 2026-09-10 at 11 51 34 AM

File: src/commands/build.rs [hunk 1]
```diff
@@ -165,136 +167,154 @@ fn build_expansion_notes(origin: &SourceOrigin, registry: &FileRegistry, notes:
...
+                if args.debug { " (debug)" } else { "" }
+            );
+            let start = Instant::now();
+            compile_assembly(&path, None, args.debug, args.arch.into())?;
+            let duration = start.elapsed();
+            println!(
+                "✅ \"{}\" built successfully in {}ms!",
+                path,
+                duration.as_micros() as f64 / 1000.0
+            );
```
Comment (bidhan-a) [inline src/commands/build.rs]: This block is repeated below on line 301 as well. Is it worth creating a small helper for this?

Comment (BretasArthur1) [reply src/commands/build.rs]: I also think it's! I'll do.

Comment (BretasArthur1) [reply src/commands/build.rs]: addressed in 5243bda

**Comment by BretasArthur1** [comment]
Screenshot 2026-09-14 at 10 49 19 PM
Screenshot 2026-09-14 at 10 49 42 PM
Screenshot 2026-09-14 at 10 50 24 PM

File: src/commands/build.rs [hunk 2]
```diff
@@ -31,8 +31,17 @@ pub struct BuildArgs {
         help = "Target architecture (v0 or v3)"
     )]
     arch: ArchArg,
-    #[arg(short = 'd', long, help = "Output deploy directory")]
+    #[arg(
+        short = 'd',
+        long,
+        help = "Output deploy directory; ignored when files are given"
```
Comment (bidhan-a) [inline src/commands/build.rs]: It might be useful to support `-d` even when explicit files are provided instead of ignoring it (unless this is intentional). This would let users choose where the generated `.so` files are written. For example:

```sh
sbpf build -d /tmp test.s
```

should write the output to `/tmp/test.so` rather than `./test.so`.

Comment (BretasArthur1) [reply src/commands/build.rs]: While designing, I decided to intentionally ignore it, so the .so is dropped where we called the cli. That's why I've also changed this: https://github.com/blueshift-gg/sbpf/pull/168/changes#diff-fecf70788e2f7552a6c0aa836593643a36d5da2acbeb4e0aad343103218668d9R222 

This was also something I've discussed with claire and she also considered to drop the file where it was called but we didn't talked about ignoring or not the flag. But now, I'm also considering that for matching behaviour, we maybe shouldn't ignore it... wdyt?

File: src/commands/build.rs [hunk 3]
```diff
@@ -165,87 +170,102 @@ fn build_expansion_notes(origin: &SourceOrigin, registry: &FileRegistry, notes:
...
+
+    // write bytecode to <filename>.so or <deploypath>/<filename>.so
+    let output_path = Path::new(deploy).join(file_name).with_extension("so");
+
+    std::fs::write(output_path, bytecode)?;
+    Ok(())
+}
+
+// Function to check if keypair file exists.
+fn has_keypair_file(dir: &Path) -> bool {
```
Comment (bidhan-a) [inline src/commands/build.rs:222]: Just curious, is a keypair generated for explicit file builds as well? Like when I run `sbpf build test.s`, will it generate both test.so and keypair.json? The keypair is only required when we run a full `sbpf build`, since it's used by the generated Mollusk test. If I just want to quickly build an assembly file with `sbpf build test.s`, then I think the keypair file is not really needed. Wdyt?

File: src/commands/build.rs [hunk 4]
```diff
@@ -262,39 +282,51 @@ pub fn build(args: BuildArgs) -> Result<()> {
...
-                let duration = start.elapsed();
-                println!(
-                    "✅ \"{}\" built successfully in {}ms!",
-                    subdir,
-                    duration.as_micros() as f64 / 1000.0
-                );
+    if is_local_build {
+        let mut outputs = HashMap::new();
+        for path in &args.path {
+            let Some(name) = Path::new(path).file_stem() else {
```
Comment (bidhan-a) [inline src/commands/build.rs]: nit:  if we have `test1/A.s` and `test2/a.s`, then it would store both `A` and `a` in the hashmap, and then output `A.so` and `a.so` to the same folder. Mac uses case-insensitive file names so I wonder how it will behave. It's an edge case so not sure if we need to worry about it.

File: src/commands/build.rs (hunk 3 above)
Comment (BretasArthur1) [reply src/commands/build.rs:222]: Yeah, I took some time to think about this, while I was building, I also considered not dropping the keypair. But while I was testing, the "maintaining behavior" pointed me to add the keypair even with a local build and the Mollusk test was actually one of the reasons why I added. My thought process was, logically seems better to not have the keypair dropped since we just want a quick way to assemble the file. But without the keypair the user wouldn't be able to quick test the file or deploy it (parts from my own testing process), and maybe this is fine it can just use its local keypair. What gave me a more objective reason to add, is when we use `sbpf build a.s` with `--deploy-dir` and suddenly we don't have the keypair there, like we have in `deploy/keypair-json` generated with just `sbpf build`, felt weird when I was testing. Overall I got to focus on maintaining behavior, and I think we don't need to consider when locally building. But when we do pass a `--deploy-dir` I think we should drop the keypair. wdyt?

File: src/commands/build.rs (hunk 4 above)
Comment (BretasArthur1) [reply src/commands/build.rs]: Dawm, nice edge case! I'll test to see how it behaves. And for a fix, I think we can get the name and convert into its raw form

Comment (bidhan-a) [reply src/commands/build.rs]: maybe an easy fix would be to store the file name in lowercase in the Hashmap (files `A` and `a` are both stored as `a`) so that the behavior is consistent across all OSs.

File: src/commands/build.rs (hunk 3 above)
Comment (bidhan-a) [reply src/commands/build.rs:222]: Yeah tbh I don't have a strong preference either. If you think there's a use case for it, I'm totally fine with generating the keypair as well.

File: src/commands/build.rs (hunk 4 above)
Comment (BretasArthur1) [reply src/commands/build.rs]: Fixed here, 4265206

File: src/commands/build.rs (hunk 3 above)
Comment (BretasArthur1) [reply src/commands/build.rs:222]: I think there's more reasons why we should've it then we shouldn't. Even that I'm not 100% sure neither.

Comment (bidhan-a) [reply src/commands/build.rs:222]: Okay, let's defer to @clairechingching on what we should do here.

**Review by bidhan-a — DISMISSED** [review]
